// vuepress-plugin-md-enhance/lib/node/markdown-it/flowchart.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.legacyFlowchart = exports.flowchart = void 0;
const hash = require("hash-sum");
const uml_1 = require("./uml");
const flowchartRender = (tokens, idx) => {
    const token = tokens[idx];
    const key = `flowchart-${hash(idx)}`;
    const { content, info } = token;
    return `<FlowChart id="${key}" code="${encodeURIComponent(content)}" preset="${info.trim().split(":", 2)[1] || "vue"}"></FlowChart>`;
};
const flowchart = (md) => {
    // Handle ```flow and ```flowchart blocks
    const fence = md.renderer.rules.fence;
    md.renderer.rules.fence = (...args) => {
        const [tokens, index] = args;
        const { info } = tokens[index];
        const realInfo = info.split(":", 2)[0];
        if (realInfo === "flow" || realInfo === "flowchart")
            return flowchartRender(tokens, index);
        return fence(...args);
    };
    md.renderer.rules["flowchart"] = flowchartRender;
};
exports.flowchart = flowchart;
/** @deprecated */
const legacyFlowchart = (md) => {
    (0, uml_1.uml)(md, {
        name: "flowchart",
        open: "flowstart",
        close: "flowend",
        render: (tokens, idx) => {
            console.warn('"@flowstart ... @flowend" is deprecated, you should use ```flow ... ``` instead.');
            const token = tokens[idx];
            const key = `flowchart_${hash(idx)}`;
            const { content, info } = token;
            return `<FlowChart id="${key}" code="${encodeURIComponent(content)}" preset="${info.trim() || "vue"}"></FlowChart>`;
        },
    });
};
exports.legacyFlowchart = legacyFlowchart;
